### ISSUE TITLE
Potential open socket leak when TLS handshake fails

### ISSUE THREAD

**Issue by vdukhovni** [vincenthz/hs-connection#27]
The `connectTo` function in `Network.Connection` calls `connectFromSocket` after opening a connection to the server:
```
connectTo :: ConnectionContext -- ^ The global context of this connection.
          -> ConnectionParams  -- ^ The parameters for this connection (where to connect, and such).
          -> IO Connection     -- ^ The new established connection on success.
connectTo cg cParams = do
    conFct <- getConFct (connectionUseSocks cParams)
    h      <- conFct (connectionHostname cParams) (N.PortNumber $ connectionPort cParams)
    connectFromSocket cg h cParams
```
The `h` handle is I believe leaked when `connectFromSocket` throws an exception.  This happens, for example, when TLS handshakes fail:
```
connectFromSocket :: ConnectionContext
                  -> Socket
                  -> ConnectionParams
                  -> IO Connection
connectFromSocket cg sock p = withSecurity (connectionUseSecure p)
    where withSecurity Nothing            = connectionNew cid $ ConnectionSocket sock
          withSecurity (Just tlsSettings) =
                tlsEstablish sock (makeTLSParams cg cid tlsSettings) >>=
                        connectionNew cid . ConnectionTLS
...
tlsEstablish :: TLS.HasBackend backend => backend -> TLS.ClientParams -> IO TLS.Context
tlsEstablish handle tlsParams = do
    ctx <- TLS.contextNew handle tlsParams
    TLS.handshake ctx
    return ctx
```
because `TLS.handshake` throws exceptions on failure and does not return the `ctx` that might otherwise be used to close the connection.

**Comment by ocheron** [comment]
So that should be resolved with connection-0.2.7.

**Comment by vincenthz** [comment]
yes indeed
